src/prefect/results.py

```diff
@@ -360,20 +369,194 @@ def key_fn():
             serialize_to_none=not self.persist_result,
         )
 
+    # TODO: These two methods need to find a new home
+
     @sync_compatible
     async def store_parameters(self, identifier: UUID, parameters: Dict[str, Any]):
-        data = self.serializer.dumps(parameters)
-        blob = PersistedResultBlob(serializer=self.serializer, data=data)
+        record = ResultRecord(
+            result=parameters,
+            metadata=ResultRecordMetadata(
+                serializer=self.serializer, storage_key=str(identifier)
+            ),
+        )
         await self.storage_block.write_path(
-            f"parameters/{identifier}", content=blob.to_bytes()
+            f"parameters/{identifier}", content=record.serialize()
         )
 
     @sync_compatible
     async def read_parameters(self, identifier: UUID) -> Dict[str, Any]:
-        blob = PersistedResultBlob.model_validate_json(
+        record = ResultRecord.deserialize(
             await self.storage_block.read_path(f"parameters/{identifier}")
         )
-        return self.serializer.loads(blob.data)
+        return record.result
+
+
+class ResultRecordMetadata(BaseModel):
+    """
+    Metadata for a result record.
+    """
+
+    storage_key: Optional[str] = Field(
+        default=None
+    )  # optional for backwards compatibility
+    expiration: Optional[DateTime] = Field(default=None)
+    serializer: Serializer = Field(default_factory=PickleSerializer)
```

Comment:
Note to self: I'd like to consider a `PrefectSerializer` that dispatches between serialization formats and optimizes for human readability (with `CloudPickle` being the ultimate fallback

Reply:
Interesting, so would it try something like JSON first and then fall back to cloudpickle?

Reply:
Yea exactly, and store the serializer it used within the metadata.